### DISCUSSION TITLE
Can i run mg_mgr_poll in the main thread and not be blocked ?

### DISCUSSION THREAD

**Discussion by gitced** [cesanta/mongoose#2147] [Ask questions, get answers] [answered]
Hello,
I'm making a project in Unreal Engine using Mongoose.
So i set up a dedicated thread for Mongoose, and have the mg_mgr_poll called in a while(true) loop.
A websockets connection is started and it all works well.
But i need to trigger some Unreal Engine actions through the websocket connection (for example loading a new map).
The problem is that Unreal Engine is checking that those actions are performed in the main thread only and crashes otherwise.
So i need Mongoose to run in the main thread, but of course i can't have a while(true) loop here.
Is there a solution to that ?
Thanks
Cedric

**Comment by scaprile** [comment, marked as answer]
I don't have the slightest idea of what Unreal Engine is.
Mongoose select()s or poll()s waiting for data or timeout (user defined, so...). If you run another stuff in the same loop, it must be non-blocking and will be delayed by Mongoose's own blocking (timeout).
If you need to communicate with other threads you can use its internal queue (latest) or its back-to-back socket connection (to be deprecated in the future) (or _a_ queue, or shared variables, or OS messages, or whatever)
https://github.com/cesanta/mongoose/tree/master/examples/multi-threaded (the example shows using Mongoose internal lock-free queue)
https://mongoose.ws/documentation/tutorials/multi-threaded/ (the tutorial covers the back-to-back socket pipe)
These options are intended to have an OS-agnostic way of doing this, and we developed the queue to be able to work with RTOSs not having support for sockets. You can use whatever interprocess mechanism you fancy and have available, of course.

**Comment by gitced** [comment]
Hello,
Thanks for your answer.
Unreal Engine is a wildly used game engine.
It uses two main threads: the game thread (used for game logic) and the render thread (used for the rendering).
And it wants everything important be called from the game thread.
Your mentionned a thread communication mechanism so i looked for something like that in Unreal and it appears UE provides various mechanisms for that.
I wasn't aware of it (i'm quite new to the world of multithreading), but your answer pushed me in the direction i needed to look for it, so, again, thanks.

For anyone bumping here and using Unreal, from a practical point of view the solution is really simple.
Just wrap your code in an "AsyncTask": choose your thread and put your code in a lambda function.
For example, to call this line in the game thread:
`gi->openCampaign(name, id);`
Simply replace it by this:
```
AsyncTask(ENamedThreads::GameThread, [gi, name, id]()
{
	gi->openCampaign(name, id);
});
```

Thanks
Cedric
